pkg/networkservice/common/excludedprefixes/server.go

```diff
@@ -0,0 +1,122 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package excludedprefixes provides a networkservice.NetworkServiceServer chain element that can read excluded prefixes
+// from config map and add them to request to avoid repeated usage.
+package excludedprefixes
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/trace"
+	"github.com/networkservicemesh/sdk/pkg/tools/prefixpool"
+)
+
+type excludedPrefixesServer struct {
+	prefixes *prefixpool.PrefixPool
+}
+
+// Note: request.Connection, Connection.Context and Context.IpContext should not be nil
+func (eps *excludedPrefixesServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	logger := trace.Log(ctx)
+
+	conn := request.Connection
+	if conn.Context.IpContext == nil {
+		conn.Context.IpContext = &networkservice.IPContext{}
+	}
+	prefixes := eps.prefixes.GetPrefixes()
+	logger.Infof("ExcludedPrefixesService: adding excluded prefixes to connection: %v", prefixes)
+	ipCtx := conn.Context.IpContext
+	ipCtx.ExcludedPrefixes = removeDuplicates(append(ipCtx.GetExcludedPrefixes(), prefixes...))
+
+	conn, err := next.Server(ctx).Request(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = eps.validateConnection(conn); err != nil {
+		logger.Errorf("ExcludedPrefixesService: connection is invalid: %v", err)
+		_, err = next.Server(ctx).Close(ctx, conn)
+		logger.Errorf("ExcludedPrefixesService: Close: %v", err)
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (eps *excludedPrefixesServer) Close(ctx context.Context, connection *networkservice.Connection) (*empty.Empty, error) {
+	return next.Server(ctx).Close(ctx, connection)
+}
+
+// NewServer -  creates a networkservice.NetworkServiceServer chain element that can read excluded prefixes from config
+// map and add them to request to avoid repeated usage.
+// Note: request.Connection, Connection.Context and Context.IpContext should not be nil when calling Request
+func NewServer() networkservice.NetworkServiceServer {
+	return NewServerFromPath(prefixpool.PrefixesFilePathDefault)
+}
+
+// NewServerFromPath -  creates a networkservice.NetworkServiceServer chain element with excluded prefixes from config map.
+func NewServerFromPath(configPath string) networkservice.NetworkServiceServer {
```

Comment:
I think we need to make this an option. Please take a look at go With...() Pattern.

Reply:
https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md

Reply:
ok, done, thank you